### DISCUSSION TITLE
How to achieve individual AMP contexts reboot？

### DISCUSSION THREAD

**Discussion by gege0500** [polarfire-soc/polarfire-soc-documentation#315] [HSS and U-Boot]
HSS v2023.02 added:
watchdog: support for rebooting individual AMP contexts
![image](https://user-images.githubusercontent.com/87520674/233917770-0b7c4c9b-e8b8-492f-8ab9-1ab0eaa556f8.png)

May I ask where the corresponding program is located in HSS? Is it here?
<img width="755" alt="0" src="https://user-images.githubusercontent.com/87520674/233919356-9465de4c-dbb6-44a9-ac8d-e68932fad6f7.png">

How can I test the individual AMP contexts reboot function?

**Comment by griffini** [comment]
Note:  in `boards/mpfs-icicle-kit/def_config`,  we have CONFIG_ALLOW_COLDREBOOT_ALWAYS` enabled.   This will make every context reboot a full chip reboot, by default. 

To test per-context reboot, ensure that `CONFIG_ALLOW_COLDREBOOT_ALWAYS` is disabled in your `.config` file in the toplevel directory. Then rebuild and reprogram the HSS to your board (`make; make program`). Issue a reboot from Linux in an AMP build, and it will reboot just the context associated with Linux. By default, a context is only allowed warm reboot itself.

The Kconfig option `CONFIG_ALLOW_COLDREBOOT`  permits cold reboots for a privileged context. In order to allow a context issue a cold reboot,  either `CONFIG_ALLOW_COLDREBOOT_ALWAYS` must enabled or the context in question must have the option the option `allow-reboot: cold` added to its entry in the `hss-payload-generator` YAML configuration file.  For example:

````
    set-name: 'PolarFire-SoC-HSS::AmpTestImage'
    hart-entry-points: {u54_1: '0x80200000', u54_2: '0x80200000', u54_3: '0x80200000, u54_4: '0xB0000000''}
    payloads:
      test/baremetal.elf: {exec-addr: '0xB0000000', owner-hart: u54_4, priv-mode: prv_m, skip-opensbi: true}
      test/u-boot-dtb.bin:    {exec-addr: '0x80200000', owner-hart: u54_1, secondary-hart: u54_2, secondary-hart: u54_3, priv-mode: prv_s, payload-name: 'u-boot-dtb', allow-reboot: cold}
````

Per-context or cold reboot is implemented by catching the `SBI_EXT_SRST` ecall.

**Reply by Shivaraj199** [reply]
hello I am running One BM Application and generating it via payload generator, I have enable cold reboot in YAML file. but the code gets hanged in the eCALL. is there any example or documentation for ecall.

YAML file
hart-entry-points: {u54_1: '0x91800000', u54_2: '0x00000000', u54_3: '0x00000000', u54_4: '0x91C00000'}
payloads:
  test/mpfs-rpmsg-master.elf: {exec-addr: '0x91800000', owner-hart: u54_1, priv-mode: prv_m, skip-opensbi: true,allow-reboot: cold}

**Reply by griffini** [reply]
Hi @Shivaraj199,

Can you provide some more details about what you are seeing? 

When you say that the "code gets hanged in the eCALL", what code is making this ecall?  If it is your mpfs-rpmsg-master.elf application, and it is using an ecall to access IHC functionality in the HSS, then you need to remove skip-opensbi: true in your code, and to make sure that you are not overriding the SBI Trap Handler.

**Reply by Shivaraj199** [reply]
Hello @griffini 

I am running a custom freeRTOS application with just standalone master context  which i hacked from amp examples running on one application core, I have removed out all the remote context hence there is no ihc. 

I am am trying to achieve the function below and  the code I tried to [refer](https://github.com/torvalds/linux/blob/master/arch/riscv/kernel/sbi.c) I pass the thwe arguments from sbi_ecall_interface.h

#define SBI_EXT_SRST_RESET			0x0

#define SBI_SRST_RESET_TYPE_SHUTDOWN		0x0
#define SBI_SRST_RESET_TYPE_COLD_REBOOT	0x1
#define SBI_SRST_RESET_TYPE_WARM_REBOOT	0x2


#define SBI_SRST_RESET_REASON_NONE	0x0
#define SBI_SRST_RESET_REASON_SYSFAIL	0x1
 sbi_ecall(int ext, int fid, unsigned long arg0,
			unsigned long arg1, unsigned long arg2,
			unsigned long arg3, unsigned long arg4,
			unsigned long arg5)
{
	struct sbiret ret;

	register uintptr_t a0 asm ("a0") = (uintptr_t)(arg0);
	register uintptr_t a1 asm ("a1") = (uintptr_t)(arg1);
	register uintptr_t a2 asm ("a2") = (uintptr_t)(arg2);
	register uintptr_t a3 asm ("a3") = (uintptr_t)(arg3);
	register uintptr_t a4 asm ("a4") = (uintptr_t)(arg4);
	register uintptr_t a5 asm ("a5") = (uintptr_t)(arg5);
	register uintptr_t a6 asm ("a6") = (uintptr_t)(fid);
	register uintptr_t a7 asm ("a7") = (uintptr_t)(ext);
	asm volatile ("ecall"
		      : "+r" (a0), "+r" (a1)
		      : "r" (a2), "r" (a3), "r" (a4), "r" (a5), "r" (a6), "r" (a7)
		      : "memory");
	ret.error = a0;
	ret.value = a1;

	return ret;
}

Am i missing something?

**Reply by griffini** [reply]
Are you running with OpenSBI (and with the standard HSS interrupt vector in place) ?

Or did you use skip-opensbi in the payload generator, and you have installed a custom interrupt handler?  If so, then you won't easily be able to use the default ecalls to perform a context reboot.
